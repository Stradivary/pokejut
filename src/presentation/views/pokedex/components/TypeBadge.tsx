import { Badge, Group, Image, Text, ThemeIcon } from '@mantine/core';

export const TypeBadge = ({ data }: { data?: { type: string; color: string; img: string; }; }) => {

    if (!data) {
        return <Text c="gray">Belum ada tipe yang dipilih</Text>;
    }

    return <Group gap={8}>
        <ThemeIcon color={data.color} c="white" size="lg" radius="xl" variant="filled">
            <Image width={32} src={data.img} alt={data.type} />
        </ThemeIcon>
        <Badge
            color={data.type === "dark" ? "gray" : data.color}
            size='lg'
            tt="capitalize"
        >
            {data.type}
        </Badge>
    </Group>;
};
