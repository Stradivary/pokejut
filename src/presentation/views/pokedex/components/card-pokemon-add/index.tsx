import { useSimulator } from "@/domain/use-cases/simulator";
import {
  Button,
  Center,
  Group,
  Image,
  Paper,
  SimpleGrid,
  Stack,
  Text,
} from "@mantine/core";
import { notifications } from "@mantine/notifications";
import { useCallback, useMemo } from "react";

import { EvolutionChain } from "@/data/entities/evolution";
import { useEvolutionChainByPokemonName } from "@/domain/use-cases/evolution";
import { usePokemonGetByName } from "@/domain/use-cases/pokemon";
import { getColorByType } from "@/utils/constants";
import { useNavigate } from "react-router-dom";
import style from "./style.module.scss";

function mapEvolutionChain(data: any): EvolutionChain {
  return {
    species: data.species.name,
    evolves_to: data.evolves_to.map((evolution: any) => mapEvolutionChain(evolution))
  };
}

export default function CardAddPokemon({
  pokemonName,
  pokemonType,
}: Readonly<{
  pokemonName: string;
  pokemonType?: any[];
}>) {
  const { catchPokemon } = useSimulator();
  const color = useMemo(() => getColorByType(pokemonType?.[0] ?? ""), [pokemonType]);

  const { data: pokemon } = usePokemonGetByName(pokemonName);

  const { data: evolveItem } = useEvolutionChainByPokemonName(pokemon?.name);

  const navigate = useNavigate();

  const selectPokemon = useCallback(() => {
    if (pokemon && evolveItem) {
      notifications.show({
        title: "Pokemon berhasil ditambahkan",
        message: `Pokemon ${pokemon.name} berhasil ditambahkan ke dalam daftar pokemon kamu`,
        color: "blue",
        icon: <img src="/pokeball.png" alt="pokeball" />,
      });
      const poke = {
        id: pokemon.id,
        name: pokemon.name,
        types: pokemon.types,
        height: pokemon.height,
        weight: pokemon.weight,
        sprites: {
          front_default: pokemon.sprites.front_default,
          back_default: "",
          other: {
            "dream_world": {
              front_default: pokemon.sprites.other["dream_world"].front_default
            },
            "home": {} as any
          }
        },
        species: pokemon.species,
      };

      const evolves_to = mapEvolutionChain(evolveItem?.chain);


      catchPokemon({ ...poke, evolves_to });
      navigate("/pokemon/selected", {
        unstable_viewTransition: true,
      });
    } else {
      notifications.show({
        title: "Gagal menambahkan pokemon",
        message: "Terdapat kesalahan saat menambahkan pokemon, silahkan coba lagi",
        color: "red",
        icon: <img src="/pokeball.png" alt="pokeball" />,
      });
    }

  }, [pokemon, evolveItem, navigate, catchPokemon]);

  return (
    <Paper
      className={style.cardAddPokemon}
      style={{
        backgroundImage: `url('/svgs/half-pokeball.svg'), radial-gradient(80% 80% at 50% bottom, ${color}, #060e20cc)`,
        backgroundRepeat: "no-repeat",
      }}
      mih={600}
      p={16}
    >
      <SimpleGrid cols={{ base: 1, xs: 1, sm: 1, md: 1, lg: 1 }} spacing={16}>
        <Image
          loading="lazy"
          style={{
            width: "100%",
            minWidth: 300,
            height: 300,
            objectFit: "contain",
          }}
          draggable={false}
          className={style.cardPokemonImg}
          src={
            pokemon?.sprites.other["dream_world"].front_default
              ? pokemon?.sprites.other["dream_world"].front_default
              : pokemon?.sprites.front_default
          }
          fallbackSrc="/pokenull.webp"
          alt={"Selected Pokemon " + pokemon?.name}
        />
        <Stack gap={16}>
          <Text className={style.pokemonName} px={20}>
            {pokemon?.name}
          </Text>
          <Group align="center" justify="center">
            {pokemon?.types?.map(
              (type: { type: { name: string; }; }, i: number) => {
                return (
                  <Image
                    loading="lazy"
                    key={`${pokemon?.name}-${type.type.name}-${i}`}
                    draggable={false}
                    w={40}
                    src={`/types/${type.type.name}.svg`}
                    alt=""
                  />
                );
              }
            )}
          </Group>
          <Group gap={16} justify="center" wrap="nowrap">
            <Group wrap="nowrap">
              <Text className={style["pokemon-stats"]}>
                {(pokemon?.height ?? 0)} M
              </Text>
              üìè
            </Group>

            <Group wrap="nowrap">
              <Text className={style["pokemon-stats"]}>
                {(pokemon?.weight ?? 0)} Kg
              </Text>
              üèãÔ∏è
            </Group>
          </Group>

        </Stack>
        <Center>
          <Button
            type="button"
            className={style["btn-choose"]}
            variant="gradient"
            onClick={selectPokemon}
            gradient={{ from: "dark", to: color ?? "blue", deg: 350 }}
          >
            I Choose You
          </Button>
        </Center>
      </SimpleGrid>
    </Paper>
  );
}
