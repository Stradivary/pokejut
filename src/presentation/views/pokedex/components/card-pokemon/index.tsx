import { Group, Image, SimpleGrid, Stack, Text } from "@mantine/core";

import React, { ReactNode, useEffect, useState } from "react";
import { getColorByType } from "@/utils/constants";
import "./style.module.scss";
import { usePokemonGetByName } from "@/domain/use-cases/pokemon";

export const CardPokemon: React.FC<{ pokemonName: string; }> = ({ pokemonName }) => {
  const [color, setColor] = useState<string | undefined>("#fff");
  const { data: pokemon } = usePokemonGetByName(pokemonName);




  useEffect(() => {
    if (pokemonName) {
      const Color = getColorByType(pokemon ? pokemon?.types[0].type.name : "");
      setColor(Color);
    }
  }, [pokemonName, pokemon]);


  useEffect(() => {
    if (pokemon) {
      document.body.classList.add("no-scroll");
    } else {
      document.body.classList.remove("no-scroll");
    }
    return () => {
      document.body.classList.remove("no-scroll");
    };
  }, [pokemon]);

  const statIcons: Record<string, ReactNode> = {
    hp: "❤️",
    attack: "⚔️",
    defense: "🛡️",
    'special-attack': "🔥",
    'special-defense': "🔥🛡️",
    speed: "🏃"
  };

  return (
    <div
      className={"card-pokemon"}
      style={{
        marginTop: 40,
        backgroundImage: `url('/svgs/half-pokeball.svg'), radial-gradient(80% 80% at 50% bottom, ${color}, #060e20cc)`,
        viewTransitionName: "pokemon-card-" + pokemonName
      }}
    >
      <Group>
        <Image
          loading="lazy"
          draggable={false}
          className="card-pokemon-img"
          style={{
            viewTransitionName: "pokemon-image-" + pokemonName
          }}
          src={
            pokemon?.sprites.other["dream_world"].front_default
              ? pokemon?.sprites.other["dream_world"].front_default
              : pokemon?.sprites.front_default
          }
          fallbackSrc="/pokenull.webp"
          alt="Selected Pokemon"
        />
        <Stack my={24}>
          <span className="card-pokemon-name">{pokemon?.name}</span>
          <Group align="center">
            {pokemon?.types?.map((type: { type: { name: string; }; }, i: number) => {
              return (
                <Image
                  loading="lazy"
                  key={i}
                  draggable={false}
                  w={40}
                  src={`/types/${type.type.name}.svg`}
                  alt=""
                />
              );
            })}
          </Group>
          <Group>
            <Group align="center">
              <Text className="pokemon-stats">
                {(pokemon?.height ?? 0)} M
              </Text>
              📏
            </Group>

            <Group align="center">
              <Text className="pokemon-stats">
                {(pokemon?.weight ?? 0)} Kg
              </Text>
              ⚖️
            </Group>
          </Group>

          <SimpleGrid cols={2}>
            {pokemon?.stats.map((stats: { base_stat: number; stat: { name: string; }; }) => {
              return (
                <Group>
                  {statIcons[stats.stat.name] ?? <></>}
                  <span>{stats.base_stat}</span>
                </Group>
              );
            })}

          </SimpleGrid>
        </Stack>

      </Group>




    </div>
  );
};
