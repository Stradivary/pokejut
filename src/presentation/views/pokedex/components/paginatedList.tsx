import React from 'react';
import CardAddPokemon from './cardAddPokemon';

export const PaginatedList = ({ pages }: { pages: any[]; }) => pages.map((page, index: number) => (
    <React.Fragment key={index + '-page'}>
        {page?.results?.map(({ name, ...rest }: { [key: string]: any; }, index: number) => (
            <CardAddPokemon
                key={name + '-card-' + index}
                pokemonName={name}
                pokemonType={rest.types} />
        ))}
    </React.Fragment>
));
