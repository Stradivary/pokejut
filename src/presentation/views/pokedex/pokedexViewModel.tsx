import { usePokemonInfiniteGetAllInternal } from "@/domain/use-cases/pokemon";
import { useIntersection } from "@mantine/hooks";
import { useEffect, useState } from "react";

const usePokedexViewModel = () => {

  const [selectedType, setSelectedType] = useState<string>("");
  const [search, setSearch] = useState<string>("");

  
  const { ref, entry } = useIntersection();

  const {
    status,
    data,
    error,
    isFetching,
    isFetchingNextPage,
    fetchNextPage,
    hasNextPage,
  } = usePokemonInfiniteGetAllInternal({ q: search, filter: selectedType });

  useEffect(() => {
    if (entry?.isIntersecting) {
      fetchNextPage();
    }
  }, [entry, fetchNextPage]);


  return {
    status,
    data,
    error,
    isFetching,
    isFetchingNextPage,
    search,
    setSearch,
    fetchNextPage,
    hasNextPage,
    intersectionRef: ref,
    selectedType,
    setSelectedType,
  };
};



export default usePokedexViewModel;
