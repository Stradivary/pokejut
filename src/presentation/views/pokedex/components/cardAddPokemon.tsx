import { useSimulator } from "@/domain/use-cases/simulator";
import {
  Button,
  Center,
  Group,
  Image,
  Paper,
  SimpleGrid,
  Stack,
  Text,
  Tooltip
} from "@mantine/core";
import { notifications } from "@mantine/notifications";
import { useCallback, useMemo } from "react";

import { addSelectedPokemon, useEvolutionChainByPokemonName } from "@/domain/use-cases/evolution";
import { usePokemonGetByName } from "@/domain/use-cases/pokemon";
import { PokemonTypeBadge } from "@/presentation/components/pokemonTypeBadge";
import { getColorByType, statLabels } from "@/utils/constants";
import { useNavigate } from "react-router-dom";
import style from "./style.module.scss";
import { getPokemonImage } from "@/utils/image";

export default function CardAddPokemon({
  pokemonName,
  pokemonType,
}: Readonly<{
  pokemonName: string;
  pokemonType?: any[];
}>) {
  const { catchPokemon } = useSimulator();
  const color = useMemo(
    () => getColorByType(pokemonType?.[0] ?? ""),
    [pokemonType]
  );
  // Informasi pokemon diambil untuk mengurangi jumlah request ke API
  // data yang disimpan disini akan digunakan untuk repository saat 
  // datanya di tambahkan ke dalam daftar pokemon
  const { data: pokemon, isSuccess: pokeDone } = usePokemonGetByName(pokemonName);
  const { data: evolveItem, isSuccess: evoDone } = useEvolutionChainByPokemonName(pokemon?.name);

  const navigate = useNavigate();

  const selectPokemon = useCallback(() => {
    const success = addSelectedPokemon(pokemon, evolveItem, catchPokemon);
    if (success) {
      notifications.show({
        title: "Pokemon berhasil ditambahkan",
        message: `Pokemon ${pokemon.name} berhasil ditambahkan ke dalam daftar pokemon kamu`,
        color: "blue",
        icon: <img src="/pokeball.png" alt="pokeball" />,
      });
      navigate("/pokemon/selected", {
        unstable_viewTransition: true,
      });
    } else {
      notifications.show({
        title: "Gagal menambahkan pokemon",
        message:
          "Terdapat kesalahan saat menambahkan pokemon, silahkan coba lagi",
        color: "red",
        icon: <img src="/pokeball.png" alt="pokeball" />,
      });
    }
    
  }, [pokemon, evolveItem, navigate, catchPokemon]);

  return (
    <Paper
      className={style.cardAddPokemon}
      style={{
        '--card-background-color': color,
      }}
      mih={600}
      p={16}
    >
      <SimpleGrid cols={{ base: 1, xs: 1, sm: 1, md: 1, lg: 1 }} spacing={16}>
        <Image
          loading="lazy"
          draggable={false}
          className={style.cardPokemonImg}
          src={getPokemonImage(pokemon)}
          fallbackSrc="/pokenull.webp"
          alt={"Selected Pokemon " + pokemon?.name}
        />
        <Stack gap={16}>
          <Text className={style.pokemonName} px={20}>
            {pokemon?.name}
          </Text>
          <Group align="center" justify="center">
            {pokemon?.types?.map(
              (type: { type: { name: string; }; }) => {
                return (
                  <PokemonTypeBadge key={`${pokemon?.name}-${type.type.name}-card`} type={type.type} />
                );
              }
            )}
          </Group>

          <Group gap={16} justify="center" wrap="nowrap">
            <Tooltip label={statLabels['height']} position="top">
              <Group wrap="nowrap" align="center">
                <Text className={style["pokemon-stats"]}>
                  {(pokemon?.height ?? 0)} M
                </Text>
                üìè
              </Group>
            </Tooltip>

            <Tooltip label={statLabels['weight']} position="top">
              <Group wrap="nowrap" align="center">
                <Text className={style["pokemon-stats"]}>
                  {(pokemon?.weight ?? 0)} Kg
                </Text>
                ‚öñÔ∏è
              </Group>
            </Tooltip>
          </Group>
        </Stack>
        <Center>
          <Button
            type="button"
            className={style["btn-choose"]}
            variant="gradient"
            disabled={!(pokeDone && evoDone)}
            onClick={selectPokemon}
            gradient={{ from: "dark", to: color ?? "blue", deg: 350 }}
          >
            I Choose You
          </Button>
        </Center>
      </SimpleGrid>
    </Paper>
  );
}
