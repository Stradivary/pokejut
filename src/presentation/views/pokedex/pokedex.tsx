// PokedexPage.js
import { pokemonData } from '@/utils/constants';
import { ActionIcon, Box, Button, Center, Flex, Group, Image, ScrollArea, SimpleGrid, Stack, Text, TextInput, Title, Tooltip, useMantineColorScheme } from '@mantine/core';
import React from 'react';
import CardAddPokemon from './components/card-pokemon-add';
import usePokedexViewModel from './pokedexViewModel'; // Import the custom hook
import { TypeBadge } from './TypeBadge';

const PokedexPage = () => {

    const { colorScheme } = useMantineColorScheme();
    const binding = usePokedexViewModel(); // Use the custom hook

    if (binding?.status === 'error') {
        return <Title>Error: {binding?.error?.message}</Title>;
    }

    return (
        <Box pos="relative" h={"calc(100vh - 32px)"}>
            <ScrollArea style={{
                height: "calc(100vh - 32px)",
                width: "100%",
            }}
                scrollbars="y"
            >

                <Flex p="md" w={{
                    base: "100%",
                    md: "calc(100% - var(--app-shell-navbar-width))"
                }} style={{
                    flexDirection: "column", position: "sticky", top: 0, left: 0, right: 0, zIndex: 10,
                    background: colorScheme === "dark" ? "#0a0a0ada" : "#ffffffda",
                    backdropFilter: "blur(8px)"
                }} >
                    <Title order={1}>Pokedex</Title>
                    <Stack w="100%">
                        <ScrollArea h={42} w="calc(100vw - 20px)"  >
                            <Group wrap='nowrap'>
                                {pokemonData?.map(({ type, color, img }: any) => {
                                    const isSelected = type === binding.selectedType;
                                    return (
                                        <Tooltip key={`action-${type}`} label={type} position="bottom" withArrow>
                                            <ActionIcon onClick={() => {
                                                if (isSelected) {
                                                    return binding.setSelectedType('');
                                                }
                                                return binding.setSelectedType(type);
                                            }} color={color} c="white" radius="xl" variant={isSelected ? "filled" : "subtle"}>
                                                <Image width={32} src={img} alt={type} />
                                            </ActionIcon>
                                        </Tooltip>
                                    );
                                })}
                            </Group>
                        </ScrollArea>
                        {
                            binding.selectedType && pokemonData?.find(({ type }: { type: string; }) => type === binding.selectedType) && (
                                <TypeBadge data={pokemonData?.find(({ type }: { type: string; }) => type === binding.selectedType)} />

                            )
                        }
                        <TextInput w="calc(100% - 32px)" value={binding.search} onChange={(x) => binding.setSearch(x.target.value)} leftSection={
                            <Image src="/search.svg" alt="search" width={20} height={20} />

                        } />
                    </Stack>
                </Flex>
                <SimpleGrid p="md"
                    w={{ base: "calc(100% - 32px)", md: "calc(100% - 300px)" }}
                    mt={32} verticalSpacing={56} spacing={12}
                    cols={{ xs: 2, sm: 3, md: 2, lg: 3, xl: 3 }}>
                    {binding?.data?.pages?.map((page: { results; }, index: number) => (
                        <React.Fragment key={index + '- page'}>
                            {page?.results?.map(({ name, ...rest }: {
                                [key: string]: any;
                            }, index: number) => (
                                <CardAddPokemon key={name + '-card-' + index}
                                    pokemonName={name}
                                    pokemonType={rest.types} />
                            ))}
                        </React.Fragment>
                    ))}

                </SimpleGrid>
                <Center mt={20}>
                    <Button
                        variant="transparent"
                        ref={binding?.intersectionRef} // Use the intersectionRef from the custom hook
                        onClick={() => binding?.fetchNextPage()}
                        loading={binding?.isFetchingNextPage}
                        disabled={!binding?.hasNextPage}
                    >
                        {
                            binding?.isFetchingNextPage
                                ? 'Menampilkan lebih banyak...'
                                : null
                        }
                        {
                            !binding?.isFetchingNextPage && binding?.hasNextPage
                                ? 'Tampilkan lebih banyak'
                                : null
                        }

                        {
                            (!binding?.hasNextPage && binding?.data?.pages?.[0] && binding?.data?.pages?.[0]?.meta?.totalPage > 0)
                                ? 'Semua Pokemon ditampilkan' : null
                        }


                    </Button>
                </Center>
                <Center>
                    <Text>{binding?.isFetching && !binding?.isFetchingNextPage ? 'Menunggu...' : null}</Text>
                </Center>
            </ScrollArea>
        </Box>
    );
};

PokedexPage.displayName = 'PokedexPage';

export const Component = PokedexPage;
