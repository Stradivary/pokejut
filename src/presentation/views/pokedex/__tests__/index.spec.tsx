import axios from "axios";
import MockAdapter from "axios-mock-adapter";
import { describe, expect, it, vi } from "vitest";
import { mockIntersectionObserver } from "~/tests/mockIntersectionObserver";
import { render } from "~/tests/test-utils";
import CardAddPokemon from "../components/card-pokemon-add";
import {
  evolveSelectedPokemon,
  mapEvolutionChain,
} from "../components/card-pokemon-add/mapEvolutionChain";
import { TypeBadge } from "../components/typeBadge";
import { Component as PokemonList } from "../pokedex";

const mock = new MockAdapter(axios);

vi.mock("Math", async (importOriginal) => {
  const m = await importOriginal<typeof Math>();
  return {
    ...m,
    random: vi.fn(() => 0.42),
  };
});

describe("Component Coverage Tests", () => {
  it("should render PokedexPage component correctly", () => {
    vi.mock("../pokedexViewModel", () => {
      return {
        default: () => {
          return {
            data: {},
          };
        },
      };
    });
    const { container } = render(<PokemonList />);
    expect(container).toMatchSnapshot();
    vi.restoreAllMocks();
  });
  it("should render PokedexPage error", () => {
    vi.mock("../pokedexViewModel", () => {
      return {
        default: () => {
          return {
            status: "error",
            error: { message: "error" },
          };
        },
      };
    });
    const { container } = render(<PokemonList />);
    expect(container).toMatchSnapshot();
    vi.restoreAllMocks();
  });
  it("should render the PokemonDetail component correctly", () => {
    mockIntersectionObserver([true]);
    const { container } = render(<PokemonList />);

    expect(container).toMatchSnapshot();
  });

  it("should render typebadge", () => {
    const { container } = render(<TypeBadge />);
    expect(container).toMatchSnapshot();
  });

  it("should render typebadge with options", () => {
    const { container } = render(
      <TypeBadge
        data={{
          color: "blue",
          img: "placeholder",
          type: "grass",
        }}
      />
    );

    expect(container).toMatchSnapshot();
  });

  it("should render the CardAddPokemon component correctly", () => {
    const mockResponse = {
      data: {
        id: 1,
        name: "pikachu",
        weight: 1,
        height: 1,
        abilities: [],
        species: { name: "", url: "" },
        sprites: {
          other: {} as any,
        },
        stats: {
          hp: 1,
          attack: 1,
          defense: 1,
          speed: 1,
        },
        types: [],
      },
    };
    mock
      .onGet("https://pokeapi.co/api/v2/pokemon/pikachu")
      .reply(200, mockResponse);
    mock.onGet("https://pokeapi.co/api/v2/pokemon-species/1").reply(200, {
      data: {
        evolution_chain: {
          url: "https://pokeapi.co/api/v2/evolution-chain/1",
        },
      },
    });
    mock
      .onGet("https://pokeapi.co/api/v2/evolution-chain/1")
      .reply(200, { data: { chain: { species: { name: "pikachu" } } } });
    const { container } = render(<CardAddPokemon pokemonName={"pikachu"} />);
    expect(container).toBeDefined();
  });
});
describe("mapEvolutionChain", () => {
  it("should map evolution chain correctly", () => {
    const data = {
      species: { name: "pikachu" },
      evolves_to: [
        {
          species: { name: "raichu" },
          evolves_to: [
            {
              species: { name: "alolan-raichu" },
              evolves_to: [],
            },
          ],
        },
      ],
    };

    const expectedEvolutionChain = {
      is_baby: false,
      species: "pikachu",
      evolves_to: [
        {
          is_baby: false,
          species: "raichu",
          evolves_to: [
            {
              is_baby: false,
              species: "alolan-raichu",
              evolves_to: [],
            },
          ],
        },
      ],
    };

    const result = mapEvolutionChain(data);

    expect(result).toEqual(expectedEvolutionChain);
  });
});

describe("evolveSelectedPokemon", () => {
  it("should return false if either pokemon or evolveItem is not provided", () => {
    const result = evolveSelectedPokemon(null, null, null);
    expect(result).toBe(false);
  });

  it("should call catchPokemon with the correct parameters", () => {
    const pokemon = {
      id: 1,
      name: "pikachu",
      types: ["electric"],
      height: 0.4,
      weight: 6,
      sprites: {
        front_default: "https://pokeapi.co/media/sprites/pokemon/25.png",
        other: {
          dream_world: {
            front_default:
              "https://pokeapi.co/media/sprites/pokemon/other/dream-world/25.svg",
          },
          home: {},
        },
      },
      species: { name: "pikachu" },
    };
    const evolveItem = {
      chain: {
        species: { name: "pikachu" },
        evolves_to: [
          {
            species: { name: "raichu" },
            evolves_to: [
              {
                species: { name: "alolan-raichu" },
                evolves_to: [],
              },
            ],
          },
        ],
      },
    };
    const catchPokemon = vi.fn();

    evolveSelectedPokemon(pokemon, evolveItem, catchPokemon);

    expect(catchPokemon).toHaveBeenCalledWith({
      id: 1,
      name: "pikachu",
      types: ["electric"],
      height: 0.4,
      weight: 6,
      sprites: {
        front_default: "https://pokeapi.co/media/sprites/pokemon/25.png",
        other: {
          dream_world: {
            front_default:
              "https://pokeapi.co/media/sprites/pokemon/other/dream-world/25.svg",
          },
          home: {},
        },
      },
      species: { name: "pikachu" },
      evolves_to: {
        is_baby: false,
        species: "pikachu",
        evolves_to: [
          {
            is_baby: false,
            species: "raichu",
            evolves_to: [
              {
                is_baby: false,
                species: "alolan-raichu",
                evolves_to: [],
              },
            ],
          },
        ],
      },
    });
  });

  it("should return true", () => {
    const pokemon = {
      id: 1,
      name: "pikachu",
      types: ["electric"],
      height: 0.4,
      weight: 6,
      sprites: {
        front_default: "https://pokeapi.co/media/sprites/pokemon/25.png",
        other: {
          dream_world: {
            front_default:
              "https://pokeapi.co/media/sprites/pokemon/other/dream-world/25.svg",
          },
          home: {},
        },
      },
      species: { name: "pikachu" },
    };
    const evolveItem = {
      chain: {
        species: { name: "pikachu" },
        evolves_to: [
          {
            species: { name: "raichu" },
            evolves_to: [
              {
                species: { name: "alolan-raichu" },
                evolves_to: [],
              },
            ],
          },
        ],
      },
    };
    const catchPokemon = vi.fn();

    const result = evolveSelectedPokemon(pokemon, evolveItem, catchPokemon);

    expect(result).toBe(true);
  });
});
