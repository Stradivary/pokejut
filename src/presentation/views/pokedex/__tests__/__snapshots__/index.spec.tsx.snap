// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Component Coverage Tests > should render PokedexPage component correctly 1`] = `
<div>
  <style
    data-mantine-styles="true"
  >
    :root{--mantine-font-family: Poppins, sans-serif;--mantine-font-family-headings: Poppins, sans-serif;--mantine-radius-default: calc(0.5rem * var(--mantine-scale));}
  </style>
  <style
    data-mantine-styles="classes"
  >
    @media (max-width: 35.99375em) {.mantine-visible-from-xs {display: none !important;}}@media (min-width: 36em) {.mantine-hidden-from-xs {display: none !important;}}@media (max-width: 47.99375em) {.mantine-visible-from-sm {display: none !important;}}@media (min-width: 48em) {.mantine-hidden-from-sm {display: none !important;}}@media (max-width: 61.99375em) {.mantine-visible-from-md {display: none !important;}}@media (min-width: 62em) {.mantine-hidden-from-md {display: none !important;}}@media (max-width: 74.99375em) {.mantine-visible-from-lg {display: none !important;}}@media (min-width: 75em) {.mantine-hidden-from-lg {display: none !important;}}@media (max-width: 87.99375em) {.mantine-visible-from-xl {display: none !important;}}@media (min-width: 88em) {.mantine-hidden-from-xl {display: none !important;}}
  </style>
  <h1
    class="m_8a5d1357 mantine-Title-root"
    data-order="1"
    style="--title-fw: var(--mantine-h1-font-weight); --title-lh: var(--mantine-h1-line-height); --title-fz: var(--mantine-h1-font-size);"
  >
    Error: 
    error
  </h1>
</div>
`;

exports[`Component Coverage Tests > should render PokedexPage error 1`] = `
<div>
  <style
    data-mantine-styles="true"
  >
    :root{--mantine-font-family: Poppins, sans-serif;--mantine-font-family-headings: Poppins, sans-serif;--mantine-radius-default: calc(0.5rem * var(--mantine-scale));}
  </style>
  <style
    data-mantine-styles="classes"
  >
    @media (max-width: 35.99375em) {.mantine-visible-from-xs {display: none !important;}}@media (min-width: 36em) {.mantine-hidden-from-xs {display: none !important;}}@media (max-width: 47.99375em) {.mantine-visible-from-sm {display: none !important;}}@media (min-width: 48em) {.mantine-hidden-from-sm {display: none !important;}}@media (max-width: 61.99375em) {.mantine-visible-from-md {display: none !important;}}@media (min-width: 62em) {.mantine-hidden-from-md {display: none !important;}}@media (max-width: 74.99375em) {.mantine-visible-from-lg {display: none !important;}}@media (min-width: 75em) {.mantine-hidden-from-lg {display: none !important;}}@media (max-width: 87.99375em) {.mantine-visible-from-xl {display: none !important;}}@media (min-width: 88em) {.mantine-hidden-from-xl {display: none !important;}}
  </style>
  <h1
    class="m_8a5d1357 mantine-Title-root"
    data-order="1"
    style="--title-fw: var(--mantine-h1-font-weight); --title-lh: var(--mantine-h1-line-height); --title-fz: var(--mantine-h1-font-size);"
  >
    Error: 
    error
  </h1>
</div>
`;

exports[`Component Coverage Tests > should render the PokemonDetail component correctly 1`] = `
<div>
  <style
    data-mantine-styles="true"
  >
    :root{--mantine-font-family: Poppins, sans-serif;--mantine-font-family-headings: Poppins, sans-serif;--mantine-radius-default: calc(0.5rem * var(--mantine-scale));}
  </style>
  <style
    data-mantine-styles="classes"
  >
    @media (max-width: 35.99375em) {.mantine-visible-from-xs {display: none !important;}}@media (min-width: 36em) {.mantine-hidden-from-xs {display: none !important;}}@media (max-width: 47.99375em) {.mantine-visible-from-sm {display: none !important;}}@media (min-width: 48em) {.mantine-hidden-from-sm {display: none !important;}}@media (max-width: 61.99375em) {.mantine-visible-from-md {display: none !important;}}@media (min-width: 62em) {.mantine-hidden-from-md {display: none !important;}}@media (max-width: 74.99375em) {.mantine-visible-from-lg {display: none !important;}}@media (min-width: 75em) {.mantine-hidden-from-lg {display: none !important;}}@media (max-width: 87.99375em) {.mantine-visible-from-xl {display: none !important;}}@media (min-width: 88em) {.mantine-hidden-from-xl {display: none !important;}}
  </style>
  <h1
    class="m_8a5d1357 mantine-Title-root"
    data-order="1"
    style="--title-fw: var(--mantine-h1-font-weight); --title-lh: var(--mantine-h1-line-height); --title-fz: var(--mantine-h1-font-size);"
  >
    Error: 
    error
  </h1>
</div>
`;

exports[`Component Coverage Tests > should render typebadge 1`] = `
<div>
  <style
    data-mantine-styles="true"
  >
    :root{--mantine-font-family: Poppins, sans-serif;--mantine-font-family-headings: Poppins, sans-serif;--mantine-radius-default: calc(0.5rem * var(--mantine-scale));}
  </style>
  <style
    data-mantine-styles="classes"
  >
    @media (max-width: 35.99375em) {.mantine-visible-from-xs {display: none !important;}}@media (min-width: 36em) {.mantine-hidden-from-xs {display: none !important;}}@media (max-width: 47.99375em) {.mantine-visible-from-sm {display: none !important;}}@media (min-width: 48em) {.mantine-hidden-from-sm {display: none !important;}}@media (max-width: 61.99375em) {.mantine-visible-from-md {display: none !important;}}@media (min-width: 62em) {.mantine-hidden-from-md {display: none !important;}}@media (max-width: 74.99375em) {.mantine-visible-from-lg {display: none !important;}}@media (min-width: 75em) {.mantine-hidden-from-lg {display: none !important;}}@media (max-width: 87.99375em) {.mantine-visible-from-xl {display: none !important;}}@media (min-width: 88em) {.mantine-hidden-from-xl {display: none !important;}}
  </style>
  <p
    class="mantine-focus-auto m_b6d8b162 mantine-Text-root"
    style="color: var(--mantine-color-gray-text);"
  >
    Belum ada tipe yang dipilih
  </p>
</div>
`;

exports[`Component Coverage Tests > should render typebadge with options 1`] = `
<div>
  <style
    data-mantine-styles="true"
  >
    :root{--mantine-font-family: Poppins, sans-serif;--mantine-font-family-headings: Poppins, sans-serif;--mantine-radius-default: calc(0.5rem * var(--mantine-scale));}
  </style>
  <style
    data-mantine-styles="classes"
  >
    @media (max-width: 35.99375em) {.mantine-visible-from-xs {display: none !important;}}@media (min-width: 36em) {.mantine-hidden-from-xs {display: none !important;}}@media (max-width: 47.99375em) {.mantine-visible-from-sm {display: none !important;}}@media (min-width: 48em) {.mantine-hidden-from-sm {display: none !important;}}@media (max-width: 61.99375em) {.mantine-visible-from-md {display: none !important;}}@media (min-width: 62em) {.mantine-hidden-from-md {display: none !important;}}@media (max-width: 74.99375em) {.mantine-visible-from-lg {display: none !important;}}@media (min-width: 75em) {.mantine-hidden-from-lg {display: none !important;}}@media (max-width: 87.99375em) {.mantine-visible-from-xl {display: none !important;}}@media (min-width: 88em) {.mantine-hidden-from-xl {display: none !important;}}
  </style>
  <div
    class="m_4081bf90 mantine-Group-root"
    style="--group-gap: calc(0.5rem * var(--mantine-scale)); --group-align: center; --group-justify: flex-start; --group-wrap: wrap;"
  >
    <div
      class="m_7341320d mantine-ThemeIcon-root"
      color="blue"
      data-size="lg"
      data-variant="filled"
      radius="xl"
      style="--ti-size: var(--ti-size-lg); --ti-radius: var(--mantine-radius-xl); --ti-bg: var(--mantine-color-blue-filled); --ti-color: var(--mantine-color-white); --ti-bd: calc(0.0625rem * var(--mantine-scale)) solid transparent; color: var(--mantine-color-white);"
    >
      <img
        alt="grass"
        class="m_9e117634 mantine-Image-root"
        src="placeholder"
        style=""
      />
    </div>
    <div
      class="m_347db0ec mantine-Badge-root"
      data-size="lg"
      style="--badge-height: var(--badge-height-lg); --badge-padding-x: var(--badge-padding-x-lg); --badge-fz: var(--badge-fz-lg); --badge-bg: var(--mantine-color-blue-filled); --badge-color: var(--mantine-color-white); --badge-bd: calc(0.0625rem * var(--mantine-scale)) solid transparent; text-transform: capitalize;"
    >
      <span
        class="m_5add502a mantine-Badge-label"
        style=""
      >
        grass
      </span>
    </div>
  </div>
</div>
`;
